lib/internal/util.js

```diff
@@ -187,3 +191,58 @@ exports.convertToValidSignal = function convertToValidSignal(signal) {
 
   throw new Error('Unknown signal: ' + signal);
 };
+
+const kCustomPromisifiedSymbol = Symbol('util.promisify.custom');
+const kCustomPromisifyArgsSymbol = Symbol('customPromisifyArgs');
+
+function promisify(orig) {
+  if (typeof orig !== 'function') {
+    throw new TypeError('The first argument to promisify() must be a function');
+  }
+
+  if (orig[kCustomPromisifiedSymbol]) {
+    return orig[kCustomPromisifiedSymbol];
+  }
+
+  if (isAsyncFunction(orig)) {
+    return orig;
+  }
+
+  // Names to create an object from in case the callback receives multiple
+  // arguments, e.g. ['stdout', 'stderr'] for child_process.exec.
+  const argumentNames = orig[kCustomPromisifyArgsSymbol];
+
+  const fn = function(...args) {
```

Comment:
I'm currently looking into adding as many tests as possible from Bluebird, one of those is the following:

```js
    specify("double promisification returns same function back", function() {
        var c = function(){};
        var a = Promise.promisify(function(){});
        var b = Promise.promisify(a);
        assert.notEqual(c, a);
        assert.strictEqual(a, b);
    });
```
Do you think it would be beneficial to keep track of promisified functions and not promisify them again?